src/couch_views/src/couch_views_indexer.erl

```diff
@@ -0,0 +1,262 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_views_indexer).
+
+-export([
+    update/2,
+    update/4,
+
+    % For tests
+    map_docs/2,
+    write_doc/4
+]).
+
+
+-include("couch_views.hrl").
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("fabric/src/fabric2.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+% TODO: 
+%  * Handle timeouts of transaction and other errors
+
+update(Db, Mrst) ->
+    Noop = fun (_) -> ok end,
+    update(Db, Mrst, Noop, []).
+
+
+update(#{} = Db, Mrst, ProgressCallback, ProgressArgs)
+        when is_function(ProgressCallback, 6) ->
+    try
+        Seq = couch_views_fdb:get_update_seq(Db, Mrst),
+        State = #{
+            since_seq => Seq,
+            count => 0,
+            limit => config:get_integer("couch_views", "change_limit", 100),
+            doc_acc => [],
+            last_seq => Seq,
+            callback => ProgressCallback,
+            callback_args => ProgressArgs,
+            mrst => Mrst
+        },
+        update_int(Db, State)
+    catch error:database_does_not_exist ->
+        #{db_prefix := DbPrefix} = Db,
+        couch_log:notice("couch_views_indexer stopped"
+        "- ~p database does not exist", [DbPrefix])
+    end.
+
+
+update_int(#{} = Db, State) ->
+    {ok, FinalState} = fabric2_fdb:transactional(Db, fun(TxDb) ->
+        State1 = maps:put(tx_db, TxDb, State),
+        fold_changes(State1)
+    end),
+
+    #{
+        count := Count,
+        limit := Limit,
+        doc_acc := DocAcc,
+        last_seq := LastSeq,
+        callback := Cb,
+        callback_args := CallbackArgs,
+        mrst := Mrst
+    } = FinalState,
+
+    {MappedResults, Mrst1} = map_docs(Mrst, DocAcc),
+    write_docs(Db, Mrst1, MappedResults, FinalState),
+
+    case Count < Limit of
+        true ->
+            Cb(undefined, finished, CallbackArgs, Db, Mrst, LastSeq);
+        false ->
+            NextState = maps:merge(FinalState, #{
+                limit => Limit,
+                count => 0,
+                doc_acc => [],
+                since_seq => LastSeq,
+                last_seq => 0,
+                mrst => Mrst1
+            }),
+            update_int(Db, NextState)
+    end.
+
+
+fold_changes(State) ->
+    #{
+        since_seq := SinceSeq,
+        limit := Limit,
+        tx_db := TxDb
+    } = State,
+
+    fabric2_db:fold_changes(TxDb, SinceSeq,
```

Comment:
this is inside a transaction. Is this a requirement?

Reply:
It's not a requirement. But it is nice to get all the docs that we want to process in 1 transaction.